### DISCUSSION TITLE
Terraform State file has the access token in clear text

### DISCUSSION THREAD

**Discussion by omerzubair** [Azure/ipam#246] [Q&A] [answered]
Hi Everyone/ @DCMattyG ,

Just some observations on using terraform the token is in clear text when i use following code. 

Anybody idea on how not to store token? My code is running on AzDo pipeline in self-hosted vmss.

data "external" "get_access_token" {
  program = ["az", "account", "get-access-token", "--resource", "api://${local.ipam_api}"]
}

![image](https://github.com/Azure/ipam/assets/7728453/602c2691-0c65-4987-89d9-93e7f4c8336f)

**Comment by DCMattyG** [comment, marked as answer]
Hi @omerzubair, I know you and I spoke about this on Teams, but just to close the loop for everyone else, this can be solved by using the Environment Variables instead of passing in the values directly as explained here:

https://registry.terraform.io/providers/XtratusCloud/azureipam/latest/docs#argument-reference

By setting the `AZUREIPAM_API_URL` and `AZUREIPAM_TOKEN` environment variables, the two values no longer need to be passed directly into the Terraform Provider and will just be read at runtime. This should also keep these values out of your state file as well.

Hope that helps other folders searching for the answer!

**Reply by omerzubair** [reply]
Yup thats correct 
following code was added in the AzDo pipeline before calling terraform apply 
```
      $ipamEngineId = "499999e-b099-7778-997e-4a*********"
      $ipamUrl = "https://myipam.azurewebsites.net"
      $ipamToken = $(az account get-access-token --resource api://$ipamEngineAppId --query accessToken --output tsv)
      $ENV:AZUREIPAM_TOKEN = $ipamToken
      $ENV:AZUREIPAM_API_URL = $ipamUrl
      
```   
And in terraform I added it like this
```
provider "azureipam" {
}
```
